Add update entity to HomeWizard

DCSBL opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add update entity with semi-hardcoded firmware versions

Needs https://github.com/home-assistant/core/pull/99598

Currently it is not possible to get any information about update from the API, but we can fake it a bit since we know the latest stable version.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/28780

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

DCSBL committed:
Add update entity for HomeWizard

DCSBL committed:
Fix copied documentation

DCSBL committed:
Add test for updateentity

DCSBL committed:
Add tests to improve coverage

DCSBL committed:
Patch version dict to allow easier updates

DCSBL committed:
Remove patch as it does not work with ============================= test session starts ==============================
platform linux -- Python 3.11.4, pytest-7.3.1, pluggy-1.0.0
rootdir: /workspaces/core
configfile: pyproject.toml
plugins: requests-mock-1.11.0, unordered-0.5.2, asyncio-0.21.0, syrupy-4.0.8, sugar-0.9.7, pytest_freezer-0.4.8, socket-0.6.0, cov-4.1.0, respx-0.20.2, xdist-3.3.1, test-groups-1.0.3, aiohttp-1.0.4, picked-0.4.6, timeout-2.1.0, anyio-3.7.1
asyncio: mode=Mode.AUTO
collected 83 items

tests/components/homewizard/test_button.py .....                         [  6%]
tests/components/homewizard/test_config_flow.py .............            [ 21%]
tests/components/homewizard/test_diagnostics.py .                        [ 22%]
tests/components/homewizard/test_init.py .......                         [ 31%]
tests/components/homewizard/test_number.py ......                        [ 38%]
tests/components/homewizard/test_sensor.py ............................. [ 73%]
..........                                                               [ 85%]
tests/components/homewizard/test_switch.py .........                     [ 96%]
tests/components/homewizard/test_update.py ...                           [100%]

=============================== warnings summary ===============================
../../usr/local/py-utils/shared/lib/python3.11/site-packages/pkg_resources/__init__.py:121
  /usr/local/py-utils/shared/lib/python3.11/site-packages/pkg_resources/__init__.py:121: DeprecationWarning: pkg_resources is deprecated as an API
    warnings.warn("pkg_resources is deprecated as an API", DeprecationWarning)

<frozen abc>:106
  <frozen abc>:106: DeprecationWarning: Inheritance class HomeAssistantApplication from web.Application is discouraged

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
======================== 83 passed, 2 warnings in 2.69s ========================

DCSBL committed:
Bumb tested version...

joostlek commented on homeassistant/components/homewizard/manifest.json:
```diff
@@ -7,6 +7,6 @@
   "loggers": ["homewizard_energy"],
   "quality_scale": "platinum",
-  "requirements": ["python-homewizard-energy==2.0.2"],
+  "requirements": ["python-homewizard-energy==2.1.0"],
```
Please create a separate PR for the dependency bump

joostlek commented on homeassistant/components/homewizard/update.py:
```diff
@@ -0,0 +1,85 @@
+
+    _attr_has_entity_name = True
+    _attr_should_poll = False
+    _attr_name = "Firmware"
```
```suggestion
```
This can be omitted as Update entities get their (translated) name from the device class.

joostlek commented on homeassistant/components/homewizard/update.py:
```diff
@@ -0,0 +1,85 @@
+        self,
+        coordinator: HWEnergyDeviceUpdateCoordinator,
+        entry: ConfigEntry,
+        release_summary: str | None = None,
```
Isn't this always None?

joostlek commented on homeassistant/components/homewizard/update.py:
```diff
@@ -0,0 +1,85 @@
+        super().__init__(coordinator)
+        self._attr_release_summary = release_summary
+        self._attr_unique_id = f"{entry.unique_id}_update"
+        self._attr_auto_update = True
+        self._attr_supported_features |= UpdateEntityFeature.RELEASE_NOTES
```
This can be moved outside of the constructor as they are static

DCSBL commented on homeassistant/components/homewizard/update.py:
```diff
@@ -0,0 +1,85 @@
+            return None
+
+    def release_notes(self) -> str | None:
+        """Return the update instructions."""
```
Not actually release notes, but it 'looks' good. We also can just add an URL to the HomeWizard support page.

DCSBL commented on tests/components/homewizard/test_update.py:
```diff
@@ -0,0 +1,96 @@
+    assert state.state == STATE_ON
+    assert state.attributes[ATTR_TITLE] is None
+    assert state.attributes[ATTR_INSTALLED_VERSION] == "1.00"
+    assert state.attributes[ATTR_LATEST_VERSION] == "1.42"
```
I tried to patch this one so we can update the version without having to adjust the test.

```
@patch("homewizard_energy.const.LATEST_FIRMWARE_VERSION", {"HWE-P1": "1.42"})
async def test_update_init(
...
```

This works when calling `python test/components/homewizard/test_update.py`, but not when calling `python test/components/homewizard` somehow.. Then the original value is used. 😕

DCSBL commented on tests/components/homewizard/test_update.py:
```diff
@@ -0,0 +1,96 @@
+    assert state.state == STATE_ON
+    assert state.attributes[ATTR_TITLE] is None
+    assert state.attributes[ATTR_INSTALLED_VERSION] == "1.00"
+    assert state.attributes[ATTR_LATEST_VERSION] == "4.18"
```
I tried to patch this one so we can update the version without having to adjust the test.

```
@patch("homewizard_energy.const.LATEST_FIRMWARE_VERSION", {"HWE-P1": "1.42"})
async def test_update_init(
...
```

This works when calling `python test/components/homewizard/test_update.py`, but not when calling `python test/components/homewizard` somehow.. Then the original value is used. 😕

joostlek replied:
try patching `homeassistant.components.homewizard.update.LATEST_FIRMWARE_VERSION`

DCSBL replied:
Yess, that works 👍

joostlek commented on homeassistant/components/homewizard/update.py:
```diff
@@ -0,0 +1,85 @@
+            return None
+
+        try:
+            return LATEST_STABLE_FIRMWARE[self.coordinator.data.device.product_type]
```
Is there a way to get the latest version from a remote service?

DCSBL replied:
No (not yet). This is the best we can provide at the moment. Feature request is added to the backlog at HomeWizard.

DCSBL committed:
Resolve PR comments

DCSBL committed:
Fix patch to allow easier updates

joostlek commented:
Did you run `script/setup`? I think you're using an older version of ruff or this PR is based on an older version of core (since it uses some old ruleset to lint the code)

DCSBL committed:
Update upstream and re-run pre-commit

DCSBL commented:
> Did you run `script/setup`? I think you're using an older version of ruff or this PR is based on an older version of core (since it uses some old ruleset to lint the code)

I was a bit behind apparently. Rebased and fixed ruff

DCSBL committed:
Merge branch 'dev' into add-update-entity

joostlek committed:
Merge branch 'dev' into add-update-entity

joostlek reviewed (changes requested):
Last remark. Since this now gets the version from the lib and has static release notes, I would like someone else to have a look if this is the way we intend to use our update entity.

DCSBL commented on homeassistant/components/homewizard/update.py:
```diff
@@ -0,0 +1,83 @@
+    _attr_auto_update = True
+    _attr_device_class = UpdateDeviceClass.FIRMWARE
+    _attr_has_entity_name = True
+    _attr_name = None
```
If I would, I get the warning that I have to set 'None' explicitly 🤷

joostlek replied:
Did that also happen after you rebased? That change was added 1 or 2 weeks ago

DCSBL replied:
Eh, before.. will take a look

DCSBL replied:
Yepp, works :)

DCSBL committed:
Remove name = None from update

DCSBL commented:
Any updates for this one?

Just checking, as I understand that draft means “no attention needed”

frenck commented:
@DCSBL I don't think we should merge this, this is not providing updates at all at this point.

I would suggest closing this, until there is an actual endpoint.

../Frenck

DCSBL commented:
I will discuss internally when we can expose a new endpoint, but this will take some time. But I agree this approach is a bit flakey.
